morf-core/src/main/java/org/alfasoftware/morf/xml/XmlDataSetConsumer.java

```diff
@@ -208,8 +213,28 @@ public void table(final Table table, final Iterable<Record> records) {
           AttributesImpl rowValueAttributes = new AttributesImpl();
           for (Column column : table.columns()) {
             String value = getValue(record, column, table.getName());
+
             if (value != null) {
-              rowValueAttributes.addAttribute(XmlDataSetNode.URI, column.getName(), column.getName(), XmlDataSetNode.STRING_TYPE, value);
+              // Escape any null characters - these aren't valid XML so in format version 3 we escape these as \0
+              // We also, consequently, have to escape backslashes as \\
+              // can't use String::replace as it doesn't like null characters
+              int last = 0;
+              StringBuilder escapedValue = new StringBuilder();

```

Comment:
You should set the initial capacity to avoid array resizing: `new StringBuilder(value.length() * 1.5)` perhaps?  i.e. 50% additional space for backslashes?

Reply:
Done